src/extension.ts

```diff
@@ -0,0 +1,35 @@
+'use strict';
+
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { getMesonTasks } from './tasks';
+
+let mesonTaskProvider: vscode.Disposable | undefined;
+
+export function activate(_context: vscode.ExtensionContext): void {
+	let workspaceRoot = vscode.workspace.rootPath;
+	if (!workspaceRoot) return;
+	
+	let pattern = path.join(workspaceRoot, 'meson.build');
+	let mesonPromise: Thenable<vscode.Task[]> | undefined = undefined;
+	let fileWatcher = vscode.workspace.createFileSystemWatcher(pattern);
+	fileWatcher.onDidChange(() => mesonPromise = undefined);
+	fileWatcher.onDidCreate(() => mesonPromise = undefined);
+	fileWatcher.onDidDelete(() => mesonPromise = undefined);
+	mesonTaskProvider = vscode.tasks.registerTaskProvider('meson', {
+		provideTasks: () => {
+			if (!mesonPromise)
+				mesonPromise = getMesonTasks(workspaceRoot);
+			return mesonPromise;
+		},
+		resolveTask(_task: vscode.Task): vscode.Task | undefined {
+			return undefined;
+		}
+	});
+}
+
+export function deactivate(): void {
+	if (mesonTaskProvider) {
+		mesonTaskProvider.dispose();
```

Comment:
you might want to set `mesonTaskProvider` to `undefined` here

Reply:
According to [vscode documentation API](https://code.visualstudio.com/docs/extensionAPI/vscode-api#Disposable), `dispose()` actually dispose the object and free resources. It should be important to clear heap memory or something.